src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,306 @@
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.PublicKeyInfo, publicKeyInfo);
+
+                    break;
+
+                default:
+                    throw new CryptographicException(SR.Format(SR.Cryptography_Cms_Invalid_Originator_Identifier_Choice, tag));
+            }
+
+            if (encodedKeyAgreement.PeekTag() == contextSpecific1)
+            {
+                // Skip extra keying material; this one isn't exposed in the API and Framwork does
+                // nothing with it.
+                encodedKeyAgreement.SkipValue();
+            }
+
+            AlgorithmIdentifier keyEncryptionAlgo = encodedKeyAgreement.ReadAlgoIdentifier();
+
+            // As there might be more than one element in RecipientEncryptedKeys, we have to create one KeyAgreeRecipientInfo for each
+            // one of the elements. 
+            List<SubjectIdentifier> recipientIdSet = new List<SubjectIdentifier>();
+            List<CryptographicAttributeObject> otherKeyMaterialSet = new List<CryptographicAttributeObject>();
+            List<DateTime> dateSet = new List<DateTime>();
+            List<byte[]> encryptionKeySet = new List<byte[]>();
+
+            ReadRecipientEncryptedKeys(encodedKeyAgreement, recipientIdSet, otherKeyMaterialSet, dateSet, encryptionKeySet);
+
+            List<KeyAgreeRecipientInfo> keyAgreeRecipients = new List<KeyAgreeRecipientInfo>(recipientIdSet.Count);
+            for (int i = 0; i < recipientIdSet.Count; i++)
+            {
+                KeyAgreeRecipientInfoPalOpenSsl kari =
+                    new KeyAgreeRecipientInfoPalOpenSsl(
+                        version,
+                        recipientIdSet[i],
+                        originatorIdentifierOrKey,
+                        otherKeyMaterialSet[i],
+                        keyEncryptionAlgo,
+                        encryptionKeySet[i],
+                        dateSet[i]);
+                keyAgreeRecipients.Add(new KeyAgreeRecipientInfo(kari));
+            }
+            return keyAgreeRecipients;
+        }
+
+        private static PublicKeyInfo ReadOriginatorPublicKey(DerSequenceReader encodedKeyAgreement)
+        {
+            // OriginatorPublicKey::= SEQUENCE {
+            //      algorithm AlgorithmIdentifier,
+            //      publicKey BIT STRING }
+            DerSequenceReader publicKeyReader = encodedKeyAgreement.ReadSequence();
+
+            DerSequenceReader algoIdentifier = publicKeyReader.ReadSequence();
+
+            Oid algoOid = algoIdentifier.ReadOid();
+            byte[] publicKey = publicKeyReader.ReadBitString();
+ 
+            // This section reads the key and determines the bit length of the key
+            byte[] rawKey = DerSequenceReader.CreateForPayload(publicKey).ReadIntegerBytes();
+            int trailingZeros = 0;
+            while (rawKey[trailingZeros++] == 0);
+            int keyLength = 8 * (rawKey.Length - trailingZeros);
```

Comment:
Are the zeros leading (which is what the code says) or trailing (which is what the variable says)?

If the key was all zeros you'd walk past the end of the array.

And this section feels like it's specific to an algorithm (presumably (classic) Diffie-Hellman) and should be guarded behind checking the OID for known values.